Support remote for Python UDF in distributed autograd

mrshenli opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#28656 Support remote for Python UDF in distributed autograd**
* #28630 Support remote for builtin operators in distributed autograd

Differential Revision: [D18138561](https://our.internmc.facebook.com/intern/diff/D18138561)

mrshenli committed:
Support remote for Python UDF in distributed autograd

[ghstack-poisoned]

mrshenli committed:
Update on "Support remote for Python UDF in distributed autograd"




[ghstack-poisoned]

zhaojuanmao commented:
overall lgtm, same comments as #28630

pritamdamania87 commented on test/dist_autograd_test.py:
```diff
@@ -433,10 +501,24 @@ def test_no_graph_with_tensors_not_require_grad(self):
+        self._test_no_graph_with_tensors_not_require_grad(ExecMode.RPC_SYNC)
 
+    @dist_init(setup_model_parallel=True)
+    def test_no_graph_with_tensors_not_require_grad_remote(self):
```
Can we test nested RPCs with RRef? For example A shares RRef to B, B shares it to C, C calls to_here() does some computation and finally the result is retrieved on A and A calls backward on this. (The RRef could live on D).

pritamdamania87 commented on torch/csrc/distributed/rpc/rref.cpp:
```diff
@@ -174,14 +174,36 @@ template <>
+    } else {
+      auto rfr = PythonRRefFetchRet::fromMessage(message);
+      future->markCompleted(c10::ivalue::Tuple::create(rfr->values()));
+    }
```
Can we dedup this code? I think we had something similar in ScriptRRefFetchRet.

mrshenli committed:
Update on "Support remote for Python UDF in distributed autograd"


Differential Revision: [D18138561](https://our.internmc.facebook.com/intern/diff/D18138561)

[ghstack-poisoned]

mrshenli committed:
Update on "Support remote for Python UDF in distributed autograd"


Differential Revision: [D18138561](https://our.internmc.facebook.com/intern/diff/D18138561)

[ghstack-poisoned]

mrshenli committed:
Update on "Support remote for Python UDF in distributed autograd"


Differential Revision: [D18138561](https://our.internmc.facebook.com/intern/diff/D18138561)

[ghstack-poisoned]

mrshenli commented:
Created https://github.com/pytorch/pytorch/issues/28819 to track the coverage on a corner case.

mrshenli committed:
Update on "Support remote for Python UDF in distributed autograd"


Differential Revision: [D18138561](https://our.internmc.facebook.com/intern/diff/D18138561)

[ghstack-poisoned]

mrshenli committed:
Update on "Support remote for Python UDF in distributed autograd"


Differential Revision: [D18138561](https://our.internmc.facebook.com/intern/diff/D18138561)

[ghstack-poisoned]

pritamdamania87 commented on test/dist_autograd_test.py:
```diff
@@ -688,6 +732,139 @@ def test_backward_simple(self):
+            futures.append(rpc.rpc_async(
+                "worker{}".format((self.rank + rank_diff) % self.world_size),
+                _run_trainer,
+                args=(rref_t1, t2, t2.grad, self_name, rank_diff)
```
Not sure I follow the test completely. It looks like the `ps` is different for each node (the ps is self.rank). Shouldn't make rank 0 the ps and ranks 1, 2 and 3 be trainers?

mrshenli replied:
This is testing 4 ps and 12 trainers concurrently. Each process x will serve as a ps and driver which will kick off 3 trainers that only talks to ps x. So we have

ps = rank0, trainer = rank1, 2, 3
ps = rank1, trainer = rank 2, 3, 4
....

Let me add some comments here to explain.

pritamdamania87 commented on test/dist_autograd_test.py:
```diff
@@ -688,6 +732,139 @@ def test_backward_simple(self):
+
+        # wait until all trainers are done
+        for fut in futures:
+            fut.wait()
```
We should validate that grads are accumulated appropriately on the PS.

pritamdamania87 commented on torch/csrc/distributed/autograd/engine/dist_engine.cpp:
```diff
@@ -246,6 +246,7 @@ void DistEngine::execute(const variable_list& roots) {
   std::shared_ptr<Node> graphRoot =
       std::make_shared<GraphRoot>(rootEdges, grads);
   edge_list outputEdges;
+
```
nit: extra empty line?

pritamdamania87 commented on torch/csrc/distributed/rpc/python_functions.cpp:
```diff
@@ -30,14 +30,14 @@ std::shared_ptr<Operator> matchBuiltinOp(
     const py::kwargs& kwargs,
     Stack& stack) {
   Symbol symbol = Symbol::fromQualString(opName);
+
```
nit: extra empty line?

pritamdamania87 commented on torch/csrc/distributed/rpc/python_functions.cpp:
```diff
@@ -163,6 +163,7 @@ std::shared_ptr<FutureMessage> pyRpcPythonUdf(
   auto pythonUDFCall = c10::guts::make_unique<PythonUDFCall>(
       std::vector<char>(pickledPythonUDF.begin(), pickledPythonUDF.end()),
       tensors);
+
```
nit: extra empty line?

mrshenli commented on test/dist_autograd_test.py:
```diff
@@ -688,6 +732,139 @@ def test_backward_simple(self):
+            else:
+                accumulate_grad_func = next_funcs[0][0]
+            self.assertEqual(
+                "torch::distributed::autograd::RecvRpcBackward",
```
good point, let me make the change.

mrshenli committed:
Update on "Support remote for Python UDF in distributed autograd"


Differential Revision: [D18138561](https://our.internmc.facebook.com/intern/diff/D18138561)

[ghstack-poisoned]

mrshenli committed:
Update on "Support remote for Python UDF in distributed autograd"


Differential Revision: [D18138561](https://our.internmc.facebook.com/intern/diff/D18138561)

[ghstack-poisoned]

mrshenli committed:
Update on "Support remote for Python UDF in distributed autograd"


Differential Revision: [D18138561](https://our.internmc.facebook.com/intern/diff/D18138561)

[ghstack-poisoned]

pritamdamania87 commented on test/dist_autograd_test.py:
```diff
@@ -69,6 +105,17 @@ def _all_contexts_cleaned_up(timeout_seconds=10):
 
+# This function creates a dis atugorad context, run rpc_sync on the given ps,
+# and then blocks until the ps has verified the grads are correctly accumulated.
+def _run_trainer(rref_t1, t2, t2_grad, ps, rank_diff):
```
t2_grad is not used here?

mrshenli committed:
Update on "Support remote for Python UDF in distributed autograd"


Differential Revision: [D18138561](https://our.internmc.facebook.com/intern/diff/D18138561)

[ghstack-poisoned]
